domain/modelagent/state/modelstate.go

```diff
@@ -1051,6 +1099,154 @@ UPDATE agent_version SET stream_id = $agentVersionStream.stream_id
 	})
 }
 
+// SetModelTargetAgentVersion is responsible for setting the current target
+// agent version of the model. This function expects a precondition version to
+// be supplied. The model's target agent version at the time the operation is
+// applied must match the preCondition version or else an error is returned.
+func (st *State) SetModelTargetAgentVersion(
+	ctx context.Context,
+	preCondition semversion.Number,
+	toVersion semversion.Number,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	checkAgentVersionStmt, err := st.Prepare(`
+SELECT &agentVersionTarget.*
+FROM   agent_version
+`,
+		agentVersionTarget{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	toVersionInput := setAgentVersionTarget{TargetVersion: toVersion.String()}
+	setAgentVersionStmt, err := st.Prepare(`
+UPDATE agent_version
+SET    target_version = $setAgentVersionTarget.target_version
+`,
+		toVersionInput,
+	)
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	preConditionVersionStr := preCondition.String()
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		currentAgentVersion := agentVersionTarget{}
+		err := tx.Query(ctx, checkAgentVersionStmt).Get(&currentAgentVersion)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return errors.New(
+				"checking current target agent version for model, no agent version has been set previously",
+			)
+		} else if err != nil {
+			return errors.Errorf(
+				"checking current target agent version for model: %w", err,
+			)
+		}
+
```

Comment:
Might be worth checking the current version doesn't already equal the target version.